File: src/lib/discovery/benchmark/validate.ts
import {
  discoveryBenchmarkRubric,
} from "./rubric";

import {
  discoveryBenchmarkScenarios,
} from "./scenarios";

import {
  getFollowUpLimit,
} from "../policies/followUp";

function assert(
  condition: boolean,
  message: string,
): void {
  if (!condition) {
    throw new Error(message);
  }
}

const totalWeight =
  Object.values(
    discoveryBenchmarkRubric,
  ).reduce(
    (total, item) =>
      total + item.weight,
    0,
  );

assert(
  totalWeight === 100,
  `Rubric must total 100. Current total: ${totalWeight}`,
);

const ids =
  discoveryBenchmarkScenarios.map(
    (scenario) => scenario.id,
  );

assert(
  new Set(ids).size === ids.length,
  "Benchmark scenario IDs must be unique.",
);

assert(
  discoveryBenchmarkScenarios.length >= 10,
  "Benchmark should contain at least 10 initial scenarios.",
);

assert(
  getFollowUpLimit(0) === 0,
  "0 answered questions should produce 0 follow-ups.",
);

assert(
  getFollowUpLimit(1) === 1,
  "1 answered question should produce 1 follow-up.",
);

assert(
  getFollowUpLimit(2) === 1,
  "2 answered questions should produce 1 follow-up.",
);

assert(
  getFollowUpLimit(3) === 2,
  "3 answered questions should produce 2 follow-ups.",
);

assert(
  getFollowUpLimit(5) === 2,
  "5 answered questions should produce 2 follow-ups.",
);

assert(
  getFollowUpLimit(10) === 2,
  "Follow-ups must never exceed 2.",
);

console.log("");
console.log("RA DISCOVERY BENCHMARK");
console.log("----------------------");
console.log(
  `Scenarios: ${discoveryBenchmarkScenarios.length}`,
);
console.log(
  `Rubric weight: ${totalWeight}/100`,
);
console.log(
  "Automatic scoring: 60 points",
);
console.log(
  "Blinded human review: 40 points",
);
console.log(
  "Follow-up policy: validated",
);
console.log("");
console.log(
  "Benchmark integrity check: PASS ✓",
);
console.log("");

File: ra-intelligence/discovery/benchmark/validate.py
from discovery.benchmark.rubric import DISCOVERY_BENCHMARK_RUBRIC
from discovery.benchmark.scenarios import DISCOVERY_BENCHMARK_SCENARIOS
from discovery.models import DiscoveryState
from discovery.policies import get_follow_up_limit


def assert_equal(actual, expected, message):
    if actual != expected:
        raise AssertionError(
            f"{message} Expected {expected!r}, received {actual!r}"
        )


def main() -> None:
    total_weight = sum(
        item["weight"]
        for item in DISCOVERY_BENCHMARK_RUBRIC.values()
    )

    assert_equal(
        total_weight,
        100,
        "Benchmark rubric must total 100.",
    )

    assert_equal(
        len(DISCOVERY_BENCHMARK_SCENARIOS),
        10,
        "Benchmark scenario count:",
    )

    scenario_ids = [
        scenario.id
        for scenario in DISCOVERY_BENCHMARK_SCENARIOS
    ]

    assert_equal(
        len(set(scenario_ids)),
        len(scenario_ids),
        "Benchmark scenario IDs must be unique.",
    )

    valid_state_fields = set(DiscoveryState.model_fields.keys())

    for scenario in DISCOVERY_BENCHMARK_SCENARIOS:
        for field_name in scenario.expected.must_capture:
            if field_name not in valid_state_fields:
                raise AssertionError(
                    f"{scenario.id}: unknown DiscoveryState field "
                    f"{field_name!r}"
                )

    assert_equal(get_follow_up_limit(0), 0, "0 questions:")
    assert_equal(get_follow_up_limit(1), 1, "1 question:")
    assert_equal(get_follow_up_limit(2), 1, "2 questions:")
    assert_equal(get_follow_up_limit(3), 2, "3 questions:")
    assert_equal(get_follow_up_limit(5), 2, "5 questions:")
    assert_equal(get_follow_up_limit(10), 2, "Follow-up cap:")

    state = DiscoveryState()

    assert_equal(
        state.service_category,
        "unknown",
        "Default service category:",
    )

    assert_equal(
        state.qualification_score,
        0,
        "Default qualification score:",
    )

    print()
    print("RA INTELLIGENCE")
    print("---------------")
    print("Python benchmark core: ONLINE")
    print(
        f"Scenarios: {len(DISCOVERY_BENCHMARK_SCENARIOS)}"
    )
    print(f"Rubric weight: {total_weight}/100")
    print("DiscoveryState fields: 20")
    print("Scenario contracts: validated")
    print("Follow-up policy: validated")
    print("Pydantic contracts: validated")
    print()
    print("Ra Intelligence integrity check: PASS ✓")
    print()


if __name__ == "__main__":
    main()
